Feat/cog visualisation epopee

MAudelGisaia opened the pull request:
- close #1043

![image](https://github.com/user-attachments/assets/53defeae-f8d7-409d-85ff-2db662233a62)

![image](https://github.com/user-attachments/assets/dc2b3355-52ec-4f36-b212-c6f71681d99a)

MAudelGisaia committed:
feat: create visualisation table
- implement interface
- save data to config
- create manage view to create data
- implement drag and drop

feat: create visualisation table
- implement interface
- save data to config
- create manage view to create data
- implement drag and drop

feat: create visualisation table
- implement interface
- save data to config
- create manage view to create data
- implement drag and drop

feat: create visualisation table
- implement interface
- save data to config
- create manage view to create data
- implement drag and drop for visualisation
- add animation

MAudelGisaia committed:
feat: finalise data group modal interface
- add condition creation
- handle edition of data groups
- create ui

MAudelGisaia committed:
fix: wrong unit

MAudelGisaia committed:
feat: register data group

MAudelGisaia committed:
feat: implement feedback

MAudelGisaia committed:
feat: implement condition visualisation and fix errors linked to this visualisation

MAudelGisaia committed:
fix: improve ui

MAudelGisaia committed:
fix: licence

MAudelGisaia committed:
feat: reorganise components
- clean log
- add comment
- refactor components

MAudelGisaia committed:
fix: add removed translation

MAudelGisaia committed:
feature: fix test and provide a way to order tab without tricks

QuCMGisaia reviewed (commented):
In the video there are a few things that I found weird:
- When you add a condition it appears at the top, which makes quite the difference in how the configuration is used, as the first condition that matches will be used for the url
- There is no information of how the conditions combine
- When editing the visualisation, the warning in the description's textarea is not related to what was being edited

MAudelGisaia commented on src/app/modules/analytics-config/components/edit-resultlist-visualisation/data-group-edition/manage-data-group-dialog.component.scss:
```diff
@@ -0,0 +1,94 @@
+    text-decoration: underline;
+    display: flex;
+    align-items: center;
+    margin-left: -9px;
```
why what ?

QuCMGisaia replied:
why the -9px?

MAudelGisaia replied:
to let less space between the icon and text

MAudelGisaia commented on src/app/modules/analytics-config/components/edit-resultlist-visualisation/manage-visualisation/manage-visualisation.component.scss:
```diff
@@ -0,0 +1,160 @@
+
+:host {
+  display: block;
+  font-size: 0.875rem;
```
Yes, if you want a fixed size for different screens, yes. But shouldn't the text be adapted to all screen sizes?

QuCMGisaia replied:
It would indeed be interesting to have that, but I think we have been removing `em` and `rem` sizes in favor of `px`. It might be something worth discussing

MAudelGisaia commented on src/app/modules/analytics-config/components/edit-resultlist-visualisation/manage-visualisation/manage-visualisation.component.scss:
```diff
@@ -0,0 +1,160 @@
+    text-decoration: underline;
+    display: flex;
+    align-items: center;
+    margin-left: -9px;
```
to let less space between the icon end the text.

MAudelGisaia committed:
fix: improve line duplication

MAudelGisaia committed:
fix:
- improve code readability
- add translation to explicit condition
- clean scss
- correct a bug that remove the values from the range input

MAudelGisaia committed:
fix:
- licence

MAudelGisaia committed:
fix:
- config

MAudelGisaia committed:
fxi: visualisation improvement.
- add more space for inputs in datagroup dialog
- update operator for condition display
- prevent missClick when user create
- visualisation for first time
- update operator to match alras api operator
- move tooltip to right

MAudelGisaia committed:
fix: error on logger

MAudelGisaia committed:
fix: add custom control initialisation for data groups that prevent the tab to be loaded

MAudelGisaia committed:
fix: improve behaviour

MAudelGisaia committed:
feat: add css class for type

MAudelGisaia committed:
fix: update css du to inputs width change. Remove duplicated component

MAudelGisaia committed:
feat: add doc, refactor filter class

QuCMGisaia commented on src/app/modules/analytics-config/components/edit-resultlist-visualisation/manage-visualisation/manage-visualisation.component.html:
```diff
@@ -0,0 +1,128 @@
+                                   [control]="visualisation()?.controls.description"></arlas-config-form-control>
+    </div>
+    <div class="body-text">
+        {{ 'Configure data eligible for visualisation' | translate: {name: (visualisation()?.value.name | lowercase)} }}
```
I still don't understand that choice, as it displays a value that is different to what the user is writing

QuCMGisaia commented on src/app/modules/analytics-config/components/edit-resultlist-visualisation/manage-visualisation/manage-visualisation.component.html:
```diff
@@ -0,0 +1,126 @@
+        </ng-container>
+        <!-- Condition Column -->
+        <ng-container matColumnDef="conditions">
+            <th mat-header-cell *matHeaderCellDef>{{ 'Condition' }}</th>
```
Missing translate pipe

QuCMGisaia replied:
Yes because if other languages are added, it can be an issue

QuCMGisaia commented on src/app/modules/analytics-config/components/edit-resultlist-visualisation/manage-visualisation/manage-visualisation.component.html:
```diff
@@ -0,0 +1,126 @@
+                    @for (filter of element.value.filters; track $index; let last = $last; let i = $index) {
+                        <li class="list-no-bullet">
+                            <span class="condition-text">{{ filter.filterField.value | translate }}</span>
+                            <span class="condition-text">{{ filter.filterOperation |lowercase | translate }}</span>
```
missing space between | and lowercase

MohamedHamouGisaia commented on src/app/modules/analytics-config/components/edit-resultlist-visualisation/result-list-visualisation.component.ts:
```diff
@@ -0,0 +1,124 @@
+    if(!this.isEdition()){
+      this.control.push(this.currentVisualisation);
+      this.currentVisualisation = null;
+      setTimeout(() => {
```
why using a settimeout ?

QuCMGisaia commented on src/app/modules/analytics-config/components/edit-resultlist-visualisation/result-list-visualisation.component.ts:
```diff
@@ -0,0 +1,124 @@
+  /** disable drag when dropping an element **/
+  public dragDisabled = true;
+  /** handle the switch between visualisation list and edit **/
+  public manageViewIsOpened = false;
```
why not use a signal?

QuCMGisaia commented on src/app/modules/analytics-config/components/edit-resultlist-visualisation/result-list-visualisation.component.ts:
```diff
@@ -0,0 +1,124 @@
+  public manageViewIsOpened = false;
+  public isEdition = signal(false);
+  /** store the current visualisation edited or created **/
+  public currentVisualisation: ResultListVisualisationsFormGroup = null;
```
why not use a signal?

MAudelGisaia replied:
Reactive is not needed. i could use it but i do not display values in the html.

QuCMGisaia commented on src/app/modules/analytics-config/services/resultlist-form-builder/models.ts:
```diff
@@ -0,0 +1,55 @@
+
+export interface ResultListVisualisationFormWidget {
+    'name': string;
+        'description': string;
```
indent is not consistent

QuCMGisaia commented on src/app/modules/analytics-config/services/resultlist-form-builder/models.ts:
```diff
@@ -0,0 +1,55 @@
+
+export  interface ResultListVisualisationConditionFormWidget {
+    filterField: {value: string;type: string;};
+    filterOperation: ArlasApiFilter ;
```
extra space before ;

QuCMGisaia commented on src/app/modules/analytics-config/services/resultlist-form-builder/resultlist-form-builder.service.ts:
```diff
@@ -701,6 +735,248 @@ export class ResultlistQuicklookFormGroup extends FormGroup {
+      name: new InputFormControl(
+        '',
+        marker('Visualisation name'),
+        marker('Name'),
```
If the description is used, the name should be less generic to avoid sharing a translation with something else

QuCMGisaia commented on src/app/modules/analytics-config/services/resultlist-form-builder/resultlist-form-builder.service.ts:
```diff
@@ -701,6 +735,248 @@ export class ResultlistQuicklookFormGroup extends FormGroup {
+      description: new TextareaFormControl(
+        '',
+        marker('Visualisation description'),
+        marker('Description'),
```
If the description is used, the name should be less generic to avoid sharing a translation with something else

QuCMGisaia replied:
The problem here is that f you use just `Description`, and somewhere else also, then you force the two places to have the same description. To solve this issue, in the rest of the builder, it was `Visualisation decription description` that was written

QuCMGisaia commented on src/app/modules/analytics-config/services/resultlist-form-builder/resultlist-form-builder.service.ts:
```diff
@@ -701,6 +735,248 @@ export class ResultlistQuicklookFormGroup extends FormGroup {
+      protocol: new SelectFormControl(
+        '',
+        marker('Protocol'),
+        marker('Protocol'),
```
If the description is used, the name should be less generic to avoid sharing a translation with something else

QuCMGisaia commented on src/app/modules/analytics-config/services/resultlist-form-builder/resultlist-form-builder.service.ts:
```diff
@@ -701,6 +735,248 @@ export class ResultlistQuicklookFormGroup extends FormGroup {
+      filterField: new TypedSelectFormControl(
+        '',
+        marker('Condition fields'),
+        marker('Condition fields'),
```
If the description is used, the name should be less generic to avoid sharing a translation with something else

QuCMGisaia commented on src/app/modules/analytics-config/services/resultlist-form-builder/resultlist-form-builder.service.ts:
```diff
@@ -701,6 +735,248 @@ export class ResultlistQuicklookFormGroup extends FormGroup {
+      ),
+      filterOperation: new SelectFormControl(
+        '',
+        marker('operation'),
```
The name should be less generic to avoid sharing a translation with something else filter operation for example)

QuCMGisaia commented on src/app/modules/analytics-config/services/resultlist-form-builder/resultlist-form-builder.service.ts:
```diff
@@ -701,6 +735,248 @@ export class ResultlistQuicklookFormGroup extends FormGroup {
+        ),
+        filterInValues: new MultipleSelectFormControl(
+          '',
+          marker('values'),
```
The name should be less generic to avoid sharing a translation with something else

QuCMGisaia commented on src/app/modules/analytics-config/services/resultlist-form-builder/resultlist-form-builder.service.ts:
```diff
@@ -701,6 +735,248 @@ export class ResultlistQuicklookFormGroup extends FormGroup {
+        ),
+        filterEqualValues: new InputFormControl(
+          '',
+          marker('values'),
```
The name should be less generic to avoid sharing a translation with something else

QuCMGisaia commented on src/app/modules/analytics-config/services/resultlist-form-builder/resultlist-form-builder.service.ts:
```diff
@@ -779,4 +1055,29 @@ export class ResultlistFormBuilderService extends WidgetFormBuilder {
+
+  public buildVisualisationsDataGroupCondition(collection: string) {
+    const collectionFields = this.collectionService.getCollectionFields(collection);
+    const operators = [ rangeArlasApiFilter,
```
inconsistent spacing in this arrayy

QuCMGisaia commented on src/app/services/main-form-manager/config-export-helper.ts:
```diff
@@ -1396,6 +1422,63 @@ export class ConfigExportHelper {
+          f.filterOperation === gtArlasApiFilter ||
+          f.filterOperation === ltArlasApiFilter ||
+          f.filterOperation === lteArlasApiFilter) &&
+      NUMERIC_TYPES.includes(condition.type as unknown as CollectionReferenceDescriptionProperty.TypeEnum)
```
if condition was typed it would simplify this line

QuCMGisaia commented on src/app/services/main-form-manager/models-config.ts:
```diff
@@ -307,12 +308,33 @@ export interface AnalyticComponentInputConfig {
+}
+
+export interface DataGroupInputCondition {
+  field: any;
```
isn't field a string?

QuCMGisaia commented on src/app/shared/models/filter-input-builder.ts:
```diff
@@ -0,0 +1,403 @@
+    this.buildInputRangeValues(parentControl, control, isLoading, METRIC_TYPES.MIN, collectionService, collection);
+  }
+
+  // eslint-disable-next-line max-len
```
is it needed?

QuCMGisaia commented on src/app/shared/models/filter-input-builder.ts:
```diff
@@ -0,0 +1,403 @@
+  public abstract keywordsFilter(parentControl: any, control: any, collectionService: CollectionService, collection: string);
+
+  /**
+   * Enable inputs if the column type is numebr and if the operator match
```
numebr -> number

QuCMGisaia commented on src/app/shared/models/filter-input-builder.ts:
```diff
@@ -0,0 +1,403 @@
+  protected abstract  getNumericalOperatorList(): { value: V; label: L; }[];
+
+  /**
+   * check if the inputs is in edit mode or not
```
if the input or is -> are

QuCMGisaia commented on src/app/shared/models/filter-input-builder.ts:
```diff
@@ -0,0 +1,403 @@
+   */
+  protected checkEditState(parentControl: ParentControl){
+    if (parentControl.editionInfo) {
+      // if we change the field/or operation, we are no longer in editing an existing filter but creating a new one
```
we are no longer editing

QuCMGisaia commented on src/app/shared/models/filter-input-builder.ts:
```diff
@@ -0,0 +1,403 @@
+
+  public manageParentControlEditState<P extends  ParentControl>(parentControl: P){
+    if (parentControl.editionInfo) {
+      // if we change the field/or operation, we are no longer in editing an existing filter but creating a new one
```
in editing -> editing

QuCMGisaia commented on src/app/shared/models/filter-input-builder.ts:
```diff
@@ -0,0 +1,403 @@
+export class FilterInputsBuilder extends InputFilter<FILTER_OPERATION, FILTER_OPERATION>{
+  protected getBooleanOperatorList(){
+    return [
+      { value: FILTER_OPERATION.IS, label: FILTER_OPERATION.IS }
```
why have a structure with value and label if both are the same?

QuCMGisaia commented on src/app/shared/models/filter-input-builder.ts:
```diff
@@ -0,0 +1,403 @@
+
+  public numberFilter<P extends  ParentControl, C extends FilterFrom>(parentControl: P, control: C) {
+    const inputValue = parentControl.customControls.filterOperation.value;
+    const enable = (inputValue === eqArlasApiFilter ||
```
I think I have seen this condition elsewhere also, maybe it could be put in a function ?

QuCMGisaia commented on src/assets/i18n/en.json:
```diff
@@ -913,6 +921,28 @@
+    "Resultlist grid": "Grid",
+	"Visualisation description": "Visualisation description",
+	"Visualisation name": "Visualisation name",
+	"Conditions will be evaluated in order" : "Conditions will be evaluated in order. The first condition to be true will be used for visualization."
```
for the visualization OR to visualize the item

QuCMGisaia commented on src/assets/i18n/fr.json:
```diff
@@ -781,6 +782,14 @@
 	"Fixed color": "Couleur unique",
 	"Creation date": "Date de création",
 	"View": "Voir",
+	"lt": "moins",
```
plus petit

QuCMGisaia commented on src/assets/i18n/fr.json:
```diff
@@ -781,6 +782,14 @@
 	"Creation date": "Date de création",
 	"View": "Voir",
+	"lt": "moins",
+	"lte": "moins ou égal à",
```
plus petit

QuCMGisaia commented on src/assets/i18n/fr.json:
```diff
@@ -912,6 +921,28 @@
+	"View URL": "URL de la visualisation",
+	"Protocol": "Protocole",
+	"Data group name": "Nom du groupe de données",
+	"Add an item family": "Ajouter une famille d'élément",
```
d'éléments

QuCMGisaia commented on src/assets/i18n/en.json:
```diff
@@ -913,6 +921,28 @@
+	"Add an item family": "Add an item family",
+	"Add a visualization configuration": "Add a visualization configuration",
+	"Add a visualization": "Add a visualization",
+	"Apply to": "Apply to",
```
`Applied to` to match the french translation

QuCMGisaia commented on src/assets/i18n/fr.json:
```diff
@@ -912,6 +921,28 @@
+	"Data group field filter": "Champ pour le filtre",
+	"Data group operator": "Opérateur",
+	"Data group field value": "Valeur de champ",
+	"Back to visualisation list": "Retour à la list des visualisations",
```
liste

QuCMGisaia reviewed (commented):
If you have the time, I suggest handling some of the quick issues that sonarqube is raising to avoid increasing the number of existing ones too much

MAudelGisaia commented on src/app/modules/analytics-config/components/edit-resultlist-visualisation/data-group-edition/manage-data-group-dialog.component.ts:
```diff
@@ -0,0 +1,132 @@
+  public ngOnInit() {
+    // init status of the button to avoid change detection errors
+    setTimeout(() => {
+      this.disableButton.set(this.data.dataGroup.invalid);
```
as disable button is a signal, the change detection ll be trigger when i set the new value.

MAudelGisaia commented on src/app/modules/analytics-config/components/edit-resultlist-visualisation/manage-visualisation/manage-visualisation.component.ts:
```diff
@@ -0,0 +1,135 @@
+    if(this.visualisation().dirty && !this.isEdition()) {
+      const confirm = this.dialog.open(ConfirmModalComponent, {
+        data: {
+          message: 'Your data has been modified do you want to leave ?'
```
translate pipe is used in

MAudelGisaia committed:
fix: sonarecube errors

QuCMGisaia commented on src/app/modules/analytics-config/components/edit-resultlist-visualisation/data-group-edition/manage-data-group-dialog.component.ts:
```diff
@@ -0,0 +1,126 @@
+
+  public ngOnInit() {
+    // init status of the button to avoid change detection errors
+    setTimeout(() => {
```
And if you trigger the change detection manually, wouldnt it work ? Or to move it in ngAfterViewInit ?

MAudelGisaia committed:
fix: implement feed

MAudelGisaia committed:
fix: change way to handle enum.

MAudelGisaia committed:
fix: remove visualisation lik, update trad.

MAudelGisaia commented on src/app/modules/analytics-config/services/resultlist-form-builder/resultlist-form-builder.service.ts:
```diff
@@ -701,6 +727,247 @@ export class ResultlistQuicklookFormGroup extends FormGroup {
+      filterField: new TypedSelectFormControl(
+        '',
+        marker('Criteria\'s fields'),
+        marker('Criteria\'s fields'),
```
its to escape quote. It will work

MAudelGisaia committed:
fix: update enum, update trad, add missing pipe.

QuCMGisaia commented on src/app/modules/analytics-config/components/edit-resultlist-visualisation/manage-visualisation/manage-visualisation.component.ts:
```diff
@@ -48,25 +48,25 @@ import { filter, first } from 'rxjs';
   protected visualisation = input<ResultListVisualisationsFormGroup>();
   /**
-   * tell if its new visualisation or not.
+   * tell Whether it is a new visualisation or not.
```
you can remove tell

QuCMGisaia commented on src/app/services/main-form-manager/config-export-helper.ts:
```diff
@@ -1448,21 +1448,24 @@ export class ConfigExportHelper {
   protected static  buildDataGroupCriteria(dataG: ResultListVisualisationDataGroupFormWidget) {
     return dataG.filters.map(f => {
+      const op = f.filterOperation as unknown as string;
+      const opArlas = (op.charAt(0).toUpperCase() + op.slice(1)) as ArlasApi;
```
It seems weird to cast to ArlasApi

QuCMGisaia commented on src/app/services/main-form-manager/models-config.ts:
```diff
@@ -327,9 +327,10 @@ export interface DataGroupInputConfig {
   filters: DataGroupInputCondition[];
 }
 
+export type ArlasApi = keyof typeof Expression.OpEnum;
```
The name could be ArlasExpression, it would be clearer

QuCMGisaia commented on src/app/utils/resultListInputsFeeder.ts:
```diff
@@ -233,6 +233,7 @@ export class ResultListInputsFeeder {
     const formArray = new FormArray([]);
     if(dataGroupConf.filters && dataGroupConf.filters.length > 0){
       dataGroupConf.filters.forEach(async (condition, i) => {
+        const op = (condition.op as string).toLowerCase() as unknown as Expression.OpEnum;
```
Here you should be able to do `Expression.OpEnum[(condition.op as string).toLowerCase()]`

MAudelGisaia commented on src/app/utils/resultListInputsFeeder.ts:
```diff
@@ -170,6 +190,125 @@ export class ResultListInputsFeeder {
     return this;
   }
 
+  public importVisualisationStep(resultListFormBuilder: ResultlistFormBuilderService){
```
@MohamedHamouGisaia @QuCMGisaia   visualisation link has been mooved to sactionStep. Do i delete this ?

MAudelGisaia committed:
fix: add visualisation link to data group (interop)

MAudelGisaia committed:
fix: translation update and css fix

MAudelGisaia committed:
fix: add default name for data group

QuCMGisaia commented on src/app/utils/resultListInputsFeeder.ts:
```diff
@@ -170,6 +181,156 @@ export class ResultListInputsFeeder {
+          control: dataGroupForm.customControls.visualisationUrl
+        }
+      ]);
+      dataGroupForm.customControls.name.setValue('Visualisation Link');
```
should be translated

QuCMGisaia commented on src/app/utils/resultListInputsFeeder.ts:
```diff
@@ -170,6 +181,156 @@ export class ResultListInputsFeeder {
+  private interopCode(resultListFormBuilder: ResultlistFormBuilderService) {
+    if(this.options.input.visualisationLink) {
+      const visualisationForm = resultListFormBuilder.buildVisualisation();
+      visualisationForm.customControls.name.setValue(' Visualisation Link');
```
extra space before Visualisation
Should be translated

QuCMGisaia commented on src/app/utils/resultListInputsFeeder.ts:
```diff
@@ -170,6 +181,156 @@ export class ResultListInputsFeeder {
+      dataGroupForm.customControls.protocol.setValue('other');
+      visualisationForm.customControls.dataGroups.push(dataGroupForm);
+
+      const key = 'Your visualisation link config has been moved';
```
Does it properly add it to the i18n files?

QuCMGisaia commented on src/app/utils/resultListInputsFeeder.ts:
```diff
@@ -37,12 +48,15 @@ export class ResultListInputsFeeder {
   protected customControls: any;
-  public constructor(protected options: ResultListConfigFeederOptions) {
+  public constructor(protected options: ResultListConfigFeederOptions, protected  messageService?: MatSnackBar,
+                     protected translate?: TranslateService) {
```
why are messageService and translate optional?

MAudelGisaia replied:
because its not really needed it juste for retrocompat code. One day it ll be removed.

QuCMGisaia commented on src/app/shared/models/filter-input-builder.ts:
```diff
@@ -0,0 +1,389 @@
+interface ParentControl {
+    editing?: boolean;
+    customControls?: any;
+    editionInfo?: { op: string | Expression.OpEnum; field: any; };
```
?

QuCMGisaia commented on src/app/shared/models/filter-input-builder.ts:
```diff
@@ -0,0 +1,389 @@
+export class GeoFilterInputsBuilder  extends InputFilter<Expression.OpEnum, Expression.OpEnum | string> {
+  protected  getBooleanOperatorList(){
+    return [
+      { value: Expression.OpEnum.Eq, label:'eq' },
```
Can you add markers to the ones which will need to be translated ? Especially is one of

MAudelGisaia committed:
fix: add new fix on translation

MAudelGisaia committed:
fix: last fixes
